File: tutorials/mpi/examples/mpi_indexed.c
/******************************************************************************
* FILE: mpi_indexed.c
* DESCRIPTION:
*   MPI tutorial example code: Indexed Derived Datatype
* AUTHOR: Blaise Barney
* LAST REVISED: 04/13/05
******************************************************************************/
#include "mpi.h"
#include <stdio.h>
#include <stdlib.h>
#define NELEMENTS 6

int main (int argc, char *argv[])
{
int numtasks, rank, source=0, dest, tag=1, i;
int blocklengths[2], displacements[2];
float a[16] =
  {1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0,
   9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0, 16.0};
float b[NELEMENTS];

MPI_Status stat;
MPI_Datatype indextype;

MPI_Init(&argc,&argv);
MPI_Comm_rank(MPI_COMM_WORLD, &rank);
MPI_Comm_size(MPI_COMM_WORLD, &numtasks);

blocklengths[0] = 4;
blocklengths[1] = 2;
displacements[0] = 5;
displacements[1] = 12;

MPI_Type_indexed(2, blocklengths, displacements, MPI_FLOAT, &indextype);
MPI_Type_commit(&indextype);

if (rank == 0) {
  for (i=0; i<numtasks; i++)
     MPI_Send(a, 1, indextype, i, tag, MPI_COMM_WORLD);
  }

MPI_Recv(b, NELEMENTS, MPI_FLOAT, source, tag, MPI_COMM_WORLD, &stat);
printf("rank= %d  b= %3.1f %3.1f %3.1f %3.1f %3.1f %3.1f\n",
      rank,b[0],b[1],b[2],b[3],b[4],b[5]);

MPI_Type_free(&indextype);
MPI_Finalize();
}

File: mpi/examples/mpi_indexed.f
C ****************************************************************************
C FILE: mpi_indexed.f
C DESCRIPTION:
C   MPI tutorial example code: Indexed Derived Datatype
C AUTHOR: Blaise Barney
C LAST REVISED: 04/02/05
C ****************************************************************************

      program indexed
      include 'mpif.h'

      integer NELEMENTS
      parameter(NELEMENTS=6)
      integer numtasks, rank, source, dest, tag, i,  ierr
      integer blocklengths(0:1), displacements(0:1)
      real*4 a(0:15), b(0:NELEMENTS-1)
      integer stat(MPI_STATUS_SIZE), indextype

      data a  /1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0,
     &         9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0, 16.0 /

      call MPI_INIT(ierr)
      call MPI_COMM_RANK(MPI_COMM_WORLD, rank, ierr)
      call MPI_COMM_SIZE(MPI_COMM_WORLD, numtasks, ierr)

      blocklengths(0) = 4
      blocklengths(1) = 2
      displacements(0) = 5
      displacements(1) = 12

      call MPI_TYPE_INDEXED(2, blocklengths, displacements, MPI_REAL, 
     &                      indextype, ierr)
      call MPI_TYPE_COMMIT(indextype, ierr)
  
      tag = 1
      if (rank .eq. 0) then
         do 10 i=0, numtasks-1
         call MPI_SEND(a, 1, indextype, i, tag, MPI_COMM_WORLD, ierr)
  10     continue
      endif

      source = 0
      call MPI_RECV(b, NELEMENTS, MPI_REAL, source, tag, MPI_COMM_WORLD, 
     &              stat, ierr)
      print *, 'rank= ',rank,' b= ',b

      call MPI_TYPE_FREE(indextype, ierr)
      call MPI_FINALIZE(ierr)

      end
